sdk/nodejs/runtime/resource.ts

```diff
@@ -227,62 +233,139 @@ async function prepareResource(label: string, res: Resource, custom: boolean,
-    const serializedProps = await serializeResourceProperties(label, props, propertyDependencies);
-
-    let parentURN: URN | undefined;
-    if (opts.parent) {
-        parentURN = await opts.parent.urn.promise();
-    } else {
-        // If no parent was provided, parent to the root resource.
-        parentURN = await getRootResource();
-    }
+    const [serializedProps, propertyToDirectDependencies] = await serializeResourceProperties(label, props);
+
+    // Wait for the parent to complete.
+    // If no parent was provided, parent to the root resource.
+    const parentURN = opts.parent
+        ? await opts.parent.urn.promise()
+        : await getRootResource();
+
+    // // Also, wait for any of our previous siblings to complete.
+    // // NOTE(cyrus): Do we really want to do this?  This is nice so that consumers don't have to
+    // // pass siblings in as dependencies of other children.  However, it limits our ability to
+    // // parallelize child resources of a component resource.
+    // if (opts.parent) {
+    //     await getTransitivelyDependendentURNs(opts.parent.__childResources, res);
+    // }
 
     let providerRef: string | undefined;
-    if (custom && (<CustomResourceOptions>opts).provider) {
-        const provider = (<CustomResourceOptions>opts).provider!;
+    if (custom && opts.provider) {
+        const provider = opts.provider;
         const providerURN = await provider.urn.promise();
         const providerID = await provider.id.promise() || unknownValue;
         providerRef = `${providerURN}::${providerID}`;
     }
 
-    const dependencies: Set<URN> = new Set<URN>(explicitDependencies);
-    const propertyDeps: Record<string, URN[]> = {};
-    for (const key of Object.keys(propertyDependencies)) {
-        const urns = new Set<URN>();
-        for (const dep of propertyDependencies[key]) {
-            const urn = await dep.urn.promise();
-            urns.add(urn);
-            dependencies.add(urn);
-        }
-        propertyDeps[key] = Array.from(urns);
+    // Collect the URNs for explicit/implicit dependencies for the engine so that it can understand
+    // the dependency graph and optimize operations accordingly.
+
+    // The list of all dependencies (implicit or explicit).
+    const allDirectDependencies = new Set<Resource>(explicitDirectDependencies);
+
+    const allDirectDependencyURNs = await getResourceURNs(explicitDirectDependencies);
+    const propertyToDirectDependencyURNs = new Map<string, Set<URN>>();
+
+    for (const [propertyName, directDependencies] of propertyToDirectDependencies) {
+        addAll(allDirectDependencies, directDependencies);
+
+        const urns = await getResourceURNs(directDependencies);
+        addAll(allDirectDependencyURNs, urns);
+        propertyToDirectDependencyURNs.set(propertyName, urns);
     }
 
+    // Now await all dependencies transitively.  i.e. if one resource depends on some other
```

Comment:
s/all dependencies transitively/the transitive closure of our dependencies' children/